nix/redis-cluster.nix

```diff
@@ -0,0 +1,145 @@
+{ pkgs, lib, name, config, ... }:
+let
+  inherit (lib) types;
+in
+{
+  options = {
+    enable = lib.mkEnableOption name;
+
+    package = lib.mkPackageOption pkgs "redis" { };
+
+    dataDir = lib.mkOption {
+      type = types.str;
+      default = "./data/${name}";
+      description = "The redis-cluster data directory (common for all nodes).";
+    };
+
+    port = lib.mkOption {
+      type = types.port;
+      default = 30001;
+      description = ''
+        The TCP port to accept connections.
+
+        Next node will use the next port and so on.
+
+        If port is set to `0`, redis will not listen on a TCP socket.
+      '';
+    };
+
+    bind = lib.mkOption {
+      type = types.nullOr types.str;
+      default = "127.0.0.1";
+      description = ''
+        The IP interface to bind to.
+        `null` means "all interfaces".
+
+        All the nodes will follow the same bind address.
+      '';
+      example = "127.0.0.1";
+    };
+
+    timeout = lib.mkOption {
+      type = types.int;
+      default = 2000;
+      description = ''
+        Time (in milliseconds) after which the node is considered to be down.
+
+        If master node is down, a replica node will take over.
+      '';
+    };
+
+    nodes = lib.mkOption {
+      type = types.int;
+      default = 6;
+      description = ''
+        Number of nodes in the cluster.
+      '';
+    };
+
+    replicas = lib.mkOption {
+      type = types.int;
+      default = 1;
+      description = ''
+        Number of replicas per Master node.
+      '';
+    };
+
+    extraConfigs = lib.mkOption {
+      type = types.attrsOf types.lines;
+      default = { };
+      description = "Attrset map of node (identified by port number) to the respective `redis.conf`.";
+      example = ''
+        {
+          {
+            30001 = "port 30001";
+          };
+        }
+      '';
+    };
+
+    outputs.settings = lib.mkOption {
+      type = types.deferredModule;
+      internal = true;
+      readOnly = true;
+      default =
+        let
+          hosts = lib.genList (x: "${config.bind}:${builtins.toString(config.port + x)}") config.nodes;
+          ports = lib.genList (x: "${builtins.toString(config.port + x)}") config.nodes;
+          healthyNodes = lib.genAttrs ports (port: { condition = "process_healthy"; });
+          node = port:
```

Comment:
Function names should be descriptive. Wouldn't something like `mkNodeProcess` be clearer?

While reading the code below I was stuck the expression `(port: node port)` and it wasn't immediately clear what it was doing, until I had to read implementation of the function being called.

Reply:
Will change the naming. I was bootstrapping a cluster that works on one node and then I generalised it to multiple nodes, but forgot to change the function name.